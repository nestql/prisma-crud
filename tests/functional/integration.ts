import "reflect-metadata";
import { promises as fs } from "fs";
import path from "path";
import util from "util";
import childProcess from "child_process";
import { buildSchema } from "type-graphql";
import { graphql } from "graphql";
import pg from "pg";

import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { getDirectoryStructureString } from "../helpers/structure";

const exec = util.promisify(childProcess.exec);

describe("generator integration", () => {
  let cwdDirPath: string;
  let schema: string;

  beforeEach(async () => {
    cwdDirPath = generateArtifactsDirPath("functional-integration");
    await fs.mkdir(cwdDirPath, { recursive: true });

    schema = /* prisma */ `
      datasource db {
        provider = "postgresql"
        url      = env("DATABASE_URL")
      }

      generator client {
        provider = "prisma-client-js"
        output   = "./generated/client"
      }

      generator typegraphql {
        provider = "node ../../../src/cli/dev.ts"
        output   = "./generated/type-graphql"
      }

      enum Color {
        RED
        GREEN
        BLUE
      }

      model User {
        id     Int      @id @default(autoincrement())
        name   String?
        posts  Post[]
      }

      model Post {
        uuid      String  @id @default(cuid())
        content   String
        author    User    @relation(fields: [authorId], references: [id])
        authorId  Int
        color     Color
      }
    `;
    await fs.writeFile(path.join(cwdDirPath, "schema.prisma"), schema);
    await fs.writeFile(
      path.join(cwdDirPath, ".env"),
      `DATABASE_URL=${process.env.TEST_DATABASE_URL}`,
    );
  });

  it("should generates TypeGraphQL classes files to output folder by running `prisma generate`", async () => {
    const prismaGenerateResult = await exec("npx prisma generate", {
      cwd: cwdDirPath,
    });
    // console.log(prismaGenerateResult);

    const directoryStructureString = getDirectoryStructureString(
      cwdDirPath + "/generated/type-graphql",
    );

    expect(prismaGenerateResult.stderr).toHaveLength(0);
    expect(directoryStructureString).toMatchSnapshot("files structure");
  }, 60000);

  it("should be able to use generate TypeGraphQL classes files to generate GraphQL schema", async () => {
    const prismaGenerateResult = await exec("npx prisma generate", {
      cwd: cwdDirPath,
    });
    // console.log(prismaGenerateResult);
    const {
      UserCrudResolver,
      PostCrudResolver,
      UserRelationsResolver,
      PostRelationsResolver,
    } = require(cwdDirPath + "/generated/type-graphql");
    await buildSchema({
      resolvers: [
        UserCrudResolver,
        PostCrudResolver,
        UserRelationsResolver,
        PostRelationsResolver,
      ],
      validate: false,
      emitSchemaFile: cwdDirPath + "/schema.graphql",
    });
    const graphQLSchemaSDL = await fs.readFile(cwdDirPath + "/schema.graphql", {
      encoding: "utf8",
    });

    expect(prismaGenerateResult.stderr).toHaveLength(0);
    expect(graphQLSchemaSDL).toMatchSnapshot("graphQLSchemaSDL");
  }, 60000);

  it("should be able to generate TypeGraphQL classes files without any type errors", async () => {
    const tsconfigContent = {
      compilerOptions: {
        target: "es2019",
        module: "commonjs",
        lib: ["es2019"],
        strict: true,
        skipLibCheck: true,
        esModuleInterop: true,
        experimentalDecorators: true,
        emitDecoratorMetadata: true,
        forceConsistentCasingInFileNames: true,
      },
    };
    const typegraphqlfolderPath = path.join(
      cwdDirPath,
      "generated",
      "type-graphql",
    );

    const prismaGenerateResult = await exec("npx prisma generate", {
      cwd: cwdDirPath,
    });
    // console.log(prismaGenerateResult);
    await fs.writeFile(
      path.join(typegraphqlfolderPath, "tsconfig.json"),
      JSON.stringify(tsconfigContent),
    );
    const tscResult = await exec("npx tsc --noEmit", {
      cwd: typegraphqlfolderPath,
    });

    expect(prismaGenerateResult.stderr).toHaveLength(0);
    expect(tscResult.stdout).toHaveLength(0);
    expect(tscResult.stderr).toHaveLength(0);
  }, 60000);

  it("should properly fetch the data from DB using PrismaClient while queried by GraphQL schema", async () => {
    const prismaGenerateResult = await exec("npx prisma generate", {
      cwd: cwdDirPath,
    });
    // console.log(prismaGenerateResult);
    expect(prismaGenerateResult.stderr).toHaveLength(0);

    // drop database before migrate
    const originalDatabaseUrl = process.env.TEST_DATABASE_URL!;
    const [dbName, ...databaseUrlParts] = originalDatabaseUrl
      .split("/")
      .reverse();
    const databaseUrl = databaseUrlParts.reverse().join("/") + "/postgres";
    const pgClient = new pg.Client({
      connectionString: databaseUrl,
    });
    await pgClient.connect();
    await pgClient.query(`DROP DATABASE IF EXISTS "${dbName}"`);
    await pgClient.end();

    const prismaMigrateResult = await exec(
      "npx prisma migrate dev --preview-feature --name init",
      { cwd: cwdDirPath },
    );
    // console.log(prismaMigrateResult);
    expect(prismaMigrateResult.stderr).toHaveLength(0);

    const { PrismaClient } = require(cwdDirPath + "/generated/client");
    const prisma = new PrismaClient();

    await prisma.user.create({ data: { name: "test1" } });
    await prisma.user.create({
      data: {
        name: "test2",
        posts: {
          create: [
            {
              color: "RED",
              content: "post content",
            },
          ],
        },
      },
    });
    await prisma.user.create({ data: { name: "not test" } });

    const {
      UserCrudResolver,
      PostCrudResolver,
      UserRelationsResolver,
      PostRelationsResolver,
    } = require(cwdDirPath + "/generated/type-graphql");
    const graphQLSchema = await buildSchema({
      resolvers: [
        UserCrudResolver,
        PostCrudResolver,
        UserRelationsResolver,
        PostRelationsResolver,
      ],
      validate: false,
    });

    const query = /* graphql */ `
      query {
        users(where: {
          name: {
            startsWith: "test"
          }
        }) {
          id
          name
          posts {
            content
            color
            author {
              name
            }
          }
        }
      }
    `;
    const { data, errors } = await graphql(graphQLSchema, query, null, {
      prisma,
    });
    await prisma.$disconnect();

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("graphql data");
  }, 100000);
});
