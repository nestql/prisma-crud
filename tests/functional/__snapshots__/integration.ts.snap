// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator integration should be able to use generate TypeGraphQL classes files to generate GraphQL schema: graphQLSchemaSDL 1`] = `
"# -----------------------------------------------
# !!! THIS FILE WAS GENERATED BY TYPE-GRAPHQL !!!
# !!!   DO NOT MODIFY THIS FILE BY YOURSELF   !!!
# -----------------------------------------------

type AffectedRowsOutput {
  count: Int!
}

type AggregatePost {
  _avg: PostAvgAggregate
  _count: PostCountAggregate
  _max: PostMaxAggregate
  _min: PostMinAggregate
  _sum: PostSumAggregate
}

type AggregateUser {
  _avg: UserAvgAggregate
  _count: UserCountAggregate
  _max: UserMaxAggregate
  _min: UserMinAggregate
  _sum: UserSumAggregate
}

enum Color {
  BLUE
  GREEN
  RED
}

input EnumColorFieldUpdateOperationsInput {
  set: Color
}

input EnumColorFilter {
  equals: Color
  in: [Color!]
  not: NestedEnumColorFilter
  notIn: [Color!]
}

input EnumColorWithAggregatesFilter {
  _count: NestedIntFilter
  _max: NestedEnumColorFilter
  _min: NestedEnumColorFilter
  equals: Color
  in: [Color!]
  not: NestedEnumColorWithAggregatesFilter
  notIn: [Color!]
}

input IntFilter {
  equals: Int
  gt: Int
  gte: Int
  in: [Int!]
  lt: Int
  lte: Int
  not: NestedIntFilter
  notIn: [Int!]
}

input IntWithAggregatesFilter {
  _avg: NestedFloatFilter
  _count: NestedIntFilter
  _max: NestedIntFilter
  _min: NestedIntFilter
  _sum: NestedIntFilter
  equals: Int
  gt: Int
  gte: Int
  in: [Int!]
  lt: Int
  lte: Int
  not: NestedIntWithAggregatesFilter
  notIn: [Int!]
}

type Mutation {
  createManyPost(data: [PostCreateManyInput!]!, skipDuplicates: Boolean): AffectedRowsOutput!
  createManyUser(data: [UserCreateManyInput!]!, skipDuplicates: Boolean): AffectedRowsOutput!
  createOnePost(data: PostCreateInput!): Post!
  createOneUser(data: UserCreateInput!): User!
  deleteManyPost(where: PostWhereInput): AffectedRowsOutput!
  deleteManyUser(where: UserWhereInput): AffectedRowsOutput!
  deleteOnePost(where: PostWhereUniqueInput!): Post
  deleteOneUser(where: UserWhereUniqueInput!): User
  updateManyPost(data: PostUpdateManyMutationInput!, where: PostWhereInput): AffectedRowsOutput!
  updateManyUser(data: UserUpdateManyMutationInput!, where: UserWhereInput): AffectedRowsOutput!
  updateOnePost(data: PostUpdateInput!, where: PostWhereUniqueInput!): Post
  updateOneUser(data: UserUpdateInput!, where: UserWhereUniqueInput!): User
  upsertOnePost(create: PostCreateInput!, update: PostUpdateInput!, where: PostWhereUniqueInput!): Post!
  upsertOneUser(create: UserCreateInput!, update: UserUpdateInput!, where: UserWhereUniqueInput!): User!
}

input NestedEnumColorFilter {
  equals: Color
  in: [Color!]
  not: NestedEnumColorFilter
  notIn: [Color!]
}

input NestedEnumColorWithAggregatesFilter {
  _count: NestedIntFilter
  _max: NestedEnumColorFilter
  _min: NestedEnumColorFilter
  equals: Color
  in: [Color!]
  not: NestedEnumColorWithAggregatesFilter
  notIn: [Color!]
}

input NestedFloatFilter {
  equals: Float
  gt: Float
  gte: Float
  in: [Float!]
  lt: Float
  lte: Float
  not: NestedFloatFilter
  notIn: [Float!]
}

input NestedIntFilter {
  equals: Int
  gt: Int
  gte: Int
  in: [Int!]
  lt: Int
  lte: Int
  not: NestedIntFilter
  notIn: [Int!]
}

input NestedIntNullableFilter {
  equals: Int
  gt: Int
  gte: Int
  in: [Int!]
  lt: Int
  lte: Int
  not: NestedIntNullableFilter
  notIn: [Int!]
}

input NestedIntWithAggregatesFilter {
  _avg: NestedFloatFilter
  _count: NestedIntFilter
  _max: NestedIntFilter
  _min: NestedIntFilter
  _sum: NestedIntFilter
  equals: Int
  gt: Int
  gte: Int
  in: [Int!]
  lt: Int
  lte: Int
  not: NestedIntWithAggregatesFilter
  notIn: [Int!]
}

input NestedStringFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  not: NestedStringFilter
  notIn: [String!]
  startsWith: String
}

input NestedStringNullableFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  not: NestedStringNullableFilter
  notIn: [String!]
  startsWith: String
}

input NestedStringNullableWithAggregatesFilter {
  _count: NestedIntNullableFilter
  _max: NestedStringNullableFilter
  _min: NestedStringNullableFilter
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  not: NestedStringNullableWithAggregatesFilter
  notIn: [String!]
  startsWith: String
}

input NestedStringWithAggregatesFilter {
  _count: NestedIntFilter
  _max: NestedStringFilter
  _min: NestedStringFilter
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  not: NestedStringWithAggregatesFilter
  notIn: [String!]
  startsWith: String
}

input NullableStringFieldUpdateOperationsInput {
  set: String
}

type Post {
  author: User!
  authorId: Int!
  color: Color!
  content: String!
  uuid: String!
}

type PostAvgAggregate {
  authorId: Float
}

input PostAvgOrderByAggregateInput {
  authorId: SortOrder
}

type PostCountAggregate {
  _all: Int!
  authorId: Int!
  color: Int!
  content: Int!
  uuid: Int!
}

input PostCountOrderByAggregateInput {
  authorId: SortOrder
  color: SortOrder
  content: SortOrder
  uuid: SortOrder
}

input PostCreateInput {
  author: UserCreateNestedOneWithoutPostsInput!
  color: Color!
  content: String!
  uuid: String
}

input PostCreateManyAuthorInput {
  color: Color!
  content: String!
  uuid: String
}

input PostCreateManyAuthorInputEnvelope {
  data: [PostCreateManyAuthorInput!]!
  skipDuplicates: Boolean
}

input PostCreateManyInput {
  authorId: Int!
  color: Color!
  content: String!
  uuid: String
}

input PostCreateNestedManyWithoutAuthorInput {
  connect: [PostWhereUniqueInput!]
  connectOrCreate: [PostCreateOrConnectWithoutAuthorInput!]
  create: [PostCreateWithoutAuthorInput!]
  createMany: PostCreateManyAuthorInputEnvelope
}

input PostCreateOrConnectWithoutAuthorInput {
  create: PostCreateWithoutAuthorInput!
  where: PostWhereUniqueInput!
}

input PostCreateWithoutAuthorInput {
  color: Color!
  content: String!
  uuid: String
}

type PostGroupBy {
  _avg: PostAvgAggregate
  _count: PostCountAggregate
  _max: PostMaxAggregate
  _min: PostMinAggregate
  _sum: PostSumAggregate
  authorId: Int!
  color: Color!
  content: String!
  uuid: String!
}

input PostListRelationFilter {
  every: PostWhereInput
  none: PostWhereInput
  some: PostWhereInput
}

type PostMaxAggregate {
  authorId: Int
  color: Color
  content: String
  uuid: String
}

input PostMaxOrderByAggregateInput {
  authorId: SortOrder
  color: SortOrder
  content: SortOrder
  uuid: SortOrder
}

type PostMinAggregate {
  authorId: Int
  color: Color
  content: String
  uuid: String
}

input PostMinOrderByAggregateInput {
  authorId: SortOrder
  color: SortOrder
  content: SortOrder
  uuid: SortOrder
}

input PostOrderByRelationAggregateInput {
  _count: SortOrder
}

input PostOrderByWithAggregationInput {
  _avg: PostAvgOrderByAggregateInput
  _count: PostCountOrderByAggregateInput
  _max: PostMaxOrderByAggregateInput
  _min: PostMinOrderByAggregateInput
  _sum: PostSumOrderByAggregateInput
  authorId: SortOrder
  color: SortOrder
  content: SortOrder
  uuid: SortOrder
}

input PostOrderByWithRelationInput {
  author: UserOrderByWithRelationInput
  authorId: SortOrder
  color: SortOrder
  content: SortOrder
  uuid: SortOrder
}

enum PostScalarFieldEnum {
  authorId
  color
  content
  uuid
}

input PostScalarWhereInput {
  AND: [PostScalarWhereInput!]
  NOT: [PostScalarWhereInput!]
  OR: [PostScalarWhereInput!]
  authorId: IntFilter
  color: EnumColorFilter
  content: StringFilter
  uuid: StringFilter
}

input PostScalarWhereWithAggregatesInput {
  AND: [PostScalarWhereWithAggregatesInput!]
  NOT: [PostScalarWhereWithAggregatesInput!]
  OR: [PostScalarWhereWithAggregatesInput!]
  authorId: IntWithAggregatesFilter
  color: EnumColorWithAggregatesFilter
  content: StringWithAggregatesFilter
  uuid: StringWithAggregatesFilter
}

type PostSumAggregate {
  authorId: Int
}

input PostSumOrderByAggregateInput {
  authorId: SortOrder
}

input PostUpdateInput {
  author: UserUpdateOneRequiredWithoutPostsNestedInput
  color: EnumColorFieldUpdateOperationsInput
  content: StringFieldUpdateOperationsInput
  uuid: StringFieldUpdateOperationsInput
}

input PostUpdateManyMutationInput {
  color: EnumColorFieldUpdateOperationsInput
  content: StringFieldUpdateOperationsInput
  uuid: StringFieldUpdateOperationsInput
}

input PostUpdateManyWithWhereWithoutAuthorInput {
  data: PostUpdateManyMutationInput!
  where: PostScalarWhereInput!
}

input PostUpdateManyWithoutAuthorNestedInput {
  connect: [PostWhereUniqueInput!]
  connectOrCreate: [PostCreateOrConnectWithoutAuthorInput!]
  create: [PostCreateWithoutAuthorInput!]
  createMany: PostCreateManyAuthorInputEnvelope
  delete: [PostWhereUniqueInput!]
  deleteMany: [PostScalarWhereInput!]
  disconnect: [PostWhereUniqueInput!]
  set: [PostWhereUniqueInput!]
  update: [PostUpdateWithWhereUniqueWithoutAuthorInput!]
  updateMany: [PostUpdateManyWithWhereWithoutAuthorInput!]
  upsert: [PostUpsertWithWhereUniqueWithoutAuthorInput!]
}

input PostUpdateWithWhereUniqueWithoutAuthorInput {
  data: PostUpdateWithoutAuthorInput!
  where: PostWhereUniqueInput!
}

input PostUpdateWithoutAuthorInput {
  color: EnumColorFieldUpdateOperationsInput
  content: StringFieldUpdateOperationsInput
  uuid: StringFieldUpdateOperationsInput
}

input PostUpsertWithWhereUniqueWithoutAuthorInput {
  create: PostCreateWithoutAuthorInput!
  update: PostUpdateWithoutAuthorInput!
  where: PostWhereUniqueInput!
}

input PostWhereInput {
  AND: [PostWhereInput!]
  NOT: [PostWhereInput!]
  OR: [PostWhereInput!]
  author: UserRelationFilter
  authorId: IntFilter
  color: EnumColorFilter
  content: StringFilter
  uuid: StringFilter
}

input PostWhereUniqueInput {
  uuid: String
}

type Query {
  aggregatePost(cursor: PostWhereUniqueInput, orderBy: [PostOrderByWithRelationInput!], skip: Int, take: Int, where: PostWhereInput): AggregatePost!
  aggregateUser(cursor: UserWhereUniqueInput, orderBy: [UserOrderByWithRelationInput!], skip: Int, take: Int, where: UserWhereInput): AggregateUser!
  findFirstPost(cursor: PostWhereUniqueInput, distinct: [PostScalarFieldEnum!], orderBy: [PostOrderByWithRelationInput!], skip: Int, take: Int, where: PostWhereInput): Post
  findFirstUser(cursor: UserWhereUniqueInput, distinct: [UserScalarFieldEnum!], orderBy: [UserOrderByWithRelationInput!], skip: Int, take: Int, where: UserWhereInput): User
  groupByPost(by: [PostScalarFieldEnum!]!, having: PostScalarWhereWithAggregatesInput, orderBy: [PostOrderByWithAggregationInput!], skip: Int, take: Int, where: PostWhereInput): [PostGroupBy!]!
  groupByUser(by: [UserScalarFieldEnum!]!, having: UserScalarWhereWithAggregatesInput, orderBy: [UserOrderByWithAggregationInput!], skip: Int, take: Int, where: UserWhereInput): [UserGroupBy!]!
  post(where: PostWhereUniqueInput!): Post
  posts(cursor: PostWhereUniqueInput, distinct: [PostScalarFieldEnum!], orderBy: [PostOrderByWithRelationInput!], skip: Int, take: Int, where: PostWhereInput): [Post!]!
  user(where: UserWhereUniqueInput!): User
  users(cursor: UserWhereUniqueInput, distinct: [UserScalarFieldEnum!], orderBy: [UserOrderByWithRelationInput!], skip: Int, take: Int, where: UserWhereInput): [User!]!
}

enum QueryMode {
  default
  insensitive
}

enum SortOrder {
  asc
  desc
}

input StringFieldUpdateOperationsInput {
  set: String
}

input StringFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  mode: QueryMode
  not: NestedStringFilter
  notIn: [String!]
  startsWith: String
}

input StringNullableFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  mode: QueryMode
  not: NestedStringNullableFilter
  notIn: [String!]
  startsWith: String
}

input StringNullableWithAggregatesFilter {
  _count: NestedIntNullableFilter
  _max: NestedStringNullableFilter
  _min: NestedStringNullableFilter
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  mode: QueryMode
  not: NestedStringNullableWithAggregatesFilter
  notIn: [String!]
  startsWith: String
}

input StringWithAggregatesFilter {
  _count: NestedIntFilter
  _max: NestedStringFilter
  _min: NestedStringFilter
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  mode: QueryMode
  not: NestedStringWithAggregatesFilter
  notIn: [String!]
  startsWith: String
}

type User {
  _count: UserCount
  id: Int!
  name: String
  posts(cursor: PostWhereUniqueInput, distinct: [PostScalarFieldEnum!], orderBy: [PostOrderByWithRelationInput!], skip: Int, take: Int, where: PostWhereInput): [Post!]!
}

type UserAvgAggregate {
  id: Float
}

input UserAvgOrderByAggregateInput {
  id: SortOrder
}

type UserCount {
  posts: Int!
}

type UserCountAggregate {
  _all: Int!
  id: Int!
  name: Int!
}

input UserCountOrderByAggregateInput {
  id: SortOrder
  name: SortOrder
}

input UserCreateInput {
  name: String
  posts: PostCreateNestedManyWithoutAuthorInput
}

input UserCreateManyInput {
  id: Int
  name: String
}

input UserCreateNestedOneWithoutPostsInput {
  connect: UserWhereUniqueInput
  connectOrCreate: UserCreateOrConnectWithoutPostsInput
  create: UserCreateWithoutPostsInput
}

input UserCreateOrConnectWithoutPostsInput {
  create: UserCreateWithoutPostsInput!
  where: UserWhereUniqueInput!
}

input UserCreateWithoutPostsInput {
  name: String
}

type UserGroupBy {
  _avg: UserAvgAggregate
  _count: UserCountAggregate
  _max: UserMaxAggregate
  _min: UserMinAggregate
  _sum: UserSumAggregate
  id: Int!
  name: String
}

type UserMaxAggregate {
  id: Int
  name: String
}

input UserMaxOrderByAggregateInput {
  id: SortOrder
  name: SortOrder
}

type UserMinAggregate {
  id: Int
  name: String
}

input UserMinOrderByAggregateInput {
  id: SortOrder
  name: SortOrder
}

input UserOrderByWithAggregationInput {
  _avg: UserAvgOrderByAggregateInput
  _count: UserCountOrderByAggregateInput
  _max: UserMaxOrderByAggregateInput
  _min: UserMinOrderByAggregateInput
  _sum: UserSumOrderByAggregateInput
  id: SortOrder
  name: SortOrder
}

input UserOrderByWithRelationInput {
  id: SortOrder
  name: SortOrder
  posts: PostOrderByRelationAggregateInput
}

input UserRelationFilter {
  is: UserWhereInput
  isNot: UserWhereInput
}

enum UserScalarFieldEnum {
  id
  name
}

input UserScalarWhereWithAggregatesInput {
  AND: [UserScalarWhereWithAggregatesInput!]
  NOT: [UserScalarWhereWithAggregatesInput!]
  OR: [UserScalarWhereWithAggregatesInput!]
  id: IntWithAggregatesFilter
  name: StringNullableWithAggregatesFilter
}

type UserSumAggregate {
  id: Int
}

input UserSumOrderByAggregateInput {
  id: SortOrder
}

input UserUpdateInput {
  name: NullableStringFieldUpdateOperationsInput
  posts: PostUpdateManyWithoutAuthorNestedInput
}

input UserUpdateManyMutationInput {
  name: NullableStringFieldUpdateOperationsInput
}

input UserUpdateOneRequiredWithoutPostsNestedInput {
  connect: UserWhereUniqueInput
  connectOrCreate: UserCreateOrConnectWithoutPostsInput
  create: UserCreateWithoutPostsInput
  update: UserUpdateWithoutPostsInput
  upsert: UserUpsertWithoutPostsInput
}

input UserUpdateWithoutPostsInput {
  name: NullableStringFieldUpdateOperationsInput
}

input UserUpsertWithoutPostsInput {
  create: UserCreateWithoutPostsInput!
  update: UserUpdateWithoutPostsInput!
}

input UserWhereInput {
  AND: [UserWhereInput!]
  NOT: [UserWhereInput!]
  OR: [UserWhereInput!]
  id: IntFilter
  name: StringNullableFilter
  posts: PostListRelationFilter
}

input UserWhereUniqueInput {
  id: Int
}
"
`;

exports[`generator integration should generates TypeGraphQL classes files to output folder by running \`prisma generate\`: files structure 1`] = `
"
[type-graphql]
  enhance.ts
  [enums]
    Color.ts
    PostScalarFieldEnum.ts
    QueryMode.ts
    SortOrder.ts
    UserScalarFieldEnum.ts
    index.ts
  helpers.ts
  index.ts
  [models]
    Post.ts
    User.ts
    index.ts
  [resolvers]
    [crud]
      [Post]
        AggregatePostResolver.ts
        CreateManyPostResolver.ts
        CreateOnePostResolver.ts
        DeleteManyPostResolver.ts
        DeleteOnePostResolver.ts
        FindFirstPostResolver.ts
        FindManyPostResolver.ts
        FindUniquePostResolver.ts
        GroupByPostResolver.ts
        PostCrudResolver.ts
        UpdateManyPostResolver.ts
        UpdateOnePostResolver.ts
        UpsertOnePostResolver.ts
        [args]
          AggregatePostArgs.ts
          CreateManyPostArgs.ts
          CreateOnePostArgs.ts
          DeleteManyPostArgs.ts
          DeleteOnePostArgs.ts
          FindFirstPostArgs.ts
          FindManyPostArgs.ts
          FindUniquePostArgs.ts
          GroupByPostArgs.ts
          UpdateManyPostArgs.ts
          UpdateOnePostArgs.ts
          UpsertOnePostArgs.ts
          index.ts
      [User]
        AggregateUserResolver.ts
        CreateManyUserResolver.ts
        CreateOneUserResolver.ts
        DeleteManyUserResolver.ts
        DeleteOneUserResolver.ts
        FindFirstUserResolver.ts
        FindManyUserResolver.ts
        FindUniqueUserResolver.ts
        GroupByUserResolver.ts
        UpdateManyUserResolver.ts
        UpdateOneUserResolver.ts
        UpsertOneUserResolver.ts
        UserCrudResolver.ts
        [args]
          AggregateUserArgs.ts
          CreateManyUserArgs.ts
          CreateOneUserArgs.ts
          DeleteManyUserArgs.ts
          DeleteOneUserArgs.ts
          FindFirstUserArgs.ts
          FindManyUserArgs.ts
          FindUniqueUserArgs.ts
          GroupByUserArgs.ts
          UpdateManyUserArgs.ts
          UpdateOneUserArgs.ts
          UpsertOneUserArgs.ts
          index.ts
      args.index.ts
      index.ts
      resolvers-actions.index.ts
      resolvers-crud.index.ts
    [inputs]
      EnumColorFieldUpdateOperationsInput.ts
      EnumColorFilter.ts
      EnumColorWithAggregatesFilter.ts
      IntFieldUpdateOperationsInput.ts
      IntFilter.ts
      IntWithAggregatesFilter.ts
      NestedEnumColorFilter.ts
      NestedEnumColorWithAggregatesFilter.ts
      NestedFloatFilter.ts
      NestedIntFilter.ts
      NestedIntNullableFilter.ts
      NestedIntWithAggregatesFilter.ts
      NestedStringFilter.ts
      NestedStringNullableFilter.ts
      NestedStringNullableWithAggregatesFilter.ts
      NestedStringWithAggregatesFilter.ts
      NullableStringFieldUpdateOperationsInput.ts
      PostAvgOrderByAggregateInput.ts
      PostCountOrderByAggregateInput.ts
      PostCreateInput.ts
      PostCreateManyAuthorInput.ts
      PostCreateManyAuthorInputEnvelope.ts
      PostCreateManyInput.ts
      PostCreateNestedManyWithoutAuthorInput.ts
      PostCreateOrConnectWithoutAuthorInput.ts
      PostCreateWithoutAuthorInput.ts
      PostListRelationFilter.ts
      PostMaxOrderByAggregateInput.ts
      PostMinOrderByAggregateInput.ts
      PostOrderByRelationAggregateInput.ts
      PostOrderByWithAggregationInput.ts
      PostOrderByWithRelationInput.ts
      PostScalarWhereInput.ts
      PostScalarWhereWithAggregatesInput.ts
      PostSumOrderByAggregateInput.ts
      PostUpdateInput.ts
      PostUpdateManyMutationInput.ts
      PostUpdateManyWithWhereWithoutAuthorInput.ts
      PostUpdateManyWithoutAuthorNestedInput.ts
      PostUpdateWithWhereUniqueWithoutAuthorInput.ts
      PostUpdateWithoutAuthorInput.ts
      PostUpsertWithWhereUniqueWithoutAuthorInput.ts
      PostWhereInput.ts
      PostWhereUniqueInput.ts
      StringFieldUpdateOperationsInput.ts
      StringFilter.ts
      StringNullableFilter.ts
      StringNullableWithAggregatesFilter.ts
      StringWithAggregatesFilter.ts
      UserAvgOrderByAggregateInput.ts
      UserCountOrderByAggregateInput.ts
      UserCreateInput.ts
      UserCreateManyInput.ts
      UserCreateNestedOneWithoutPostsInput.ts
      UserCreateOrConnectWithoutPostsInput.ts
      UserCreateWithoutPostsInput.ts
      UserMaxOrderByAggregateInput.ts
      UserMinOrderByAggregateInput.ts
      UserOrderByWithAggregationInput.ts
      UserOrderByWithRelationInput.ts
      UserRelationFilter.ts
      UserScalarWhereWithAggregatesInput.ts
      UserSumOrderByAggregateInput.ts
      UserUpdateInput.ts
      UserUpdateManyMutationInput.ts
      UserUpdateOneRequiredWithoutPostsNestedInput.ts
      UserUpdateWithoutPostsInput.ts
      UserUpsertWithoutPostsInput.ts
      UserWhereInput.ts
      UserWhereUniqueInput.ts
      index.ts
    [outputs]
      AffectedRowsOutput.ts
      AggregatePost.ts
      AggregateUser.ts
      PostAvgAggregate.ts
      PostCountAggregate.ts
      PostGroupBy.ts
      PostMaxAggregate.ts
      PostMinAggregate.ts
      PostSumAggregate.ts
      UserAvgAggregate.ts
      UserCount.ts
      UserCountAggregate.ts
      UserGroupBy.ts
      UserMaxAggregate.ts
      UserMinAggregate.ts
      UserSumAggregate.ts
      index.ts
    [relations]
      [Post]
        PostRelationsResolver.ts
      [User]
        UserRelationsResolver.ts
        [args]
          UserPostsArgs.ts
          index.ts
      args.index.ts
      index.ts
      resolvers.index.ts
  scalars.ts
"
`;

exports[`generator integration should properly fetch the data from DB using PrismaClient while queried by GraphQL schema: graphql data 1`] = `
Object {
  "users": Array [
    Object {
      "id": 1,
      "name": "test1",
      "posts": Array [],
    },
    Object {
      "id": 2,
      "name": "test2",
      "posts": Array [
        Object {
          "author": Object {
            "name": "test2",
          },
          "color": "RED",
          "content": "post content",
        },
      ],
    },
  ],
}
`;
